File: src/main/java/uk/yermak/audiobookconverter/ArtWorkProxy.java
package uk.yermak.audiobookconverter;

import javafx.scene.image.Image;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.lang.invoke.MethodHandles;
import java.util.concurrent.Future;

public class ArtWorkProxy implements ArtWork {
    private final Future<ArtWork> futureLoad;
    private final Logger logger;

    private Logger logger() {
        return this.logger;
    }

    private ArtWork getArtWork() {
        try {
            return this.futureLoad.get();
        } catch (Exception e) {
            this.logger().error("Failed to load ArtWork Proxy:", e);
            e.printStackTrace();
            throw new RuntimeException(e);
        }
    }

    @Override
    public boolean matchCrc32(long crc32) {
        return this.getArtWork().matchCrc32(crc32);
    }

    @Override
    public Image image() {
        return getArtWork().image();
    }

    public long getCrc32() {
        return this.getArtWork().getCrc32();
    }

    public String getFileName() {
        return this.getArtWork().getFileName();
    }

    public ArtWorkProxy(final Future<ArtWork> futureLoad) {
        this.futureLoad = futureLoad;
        this.logger = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
    }
}

        

File: src/main/scala/uk/yermak/audiobookconverter/ArtWorkProxy.scala
package uk.yermak.audiobookconverter

import java.lang.invoke.MethodHandles
import java.util.concurrent.{ExecutionException, Future}

import org.slf4j.{Logger, LoggerFactory}

/**
  * Created by yermak on 1/11/2018.
  */
class ArtWorkProxy(val futureLoad: Future[ArtWork], var format: String) extends ArtWork {
  private val logger: Logger = LoggerFactory.getLogger(MethodHandles.lookup.lookupClass)

  private def getArtWork = try
    futureLoad.get
  catch {
    case e@(_: InterruptedException | _: ExecutionException) =>
      logger.error("Failed to load ArtWork Proxy:", e)
      e.printStackTrace()
      throw new RuntimeException(e)
  }

  override def getFormat: String = format

  override def setFormat(format: String): Unit = getArtWork.setFormat(format)

  override def getCrc32: Long = getArtWork.getCrc32

  override def setCrc32(crc32: Long): Unit = getArtWork.setCrc32(crc32)

  override def getFileName: String = getArtWork.getFileName

  override def setFileName(fileName: String): Unit = getArtWork.setFileName(fileName)
}